### PR TITLE
config: Fix test for _Float16

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Fixes a build issue with clang 6, which accepts the type but fails
when used in arithmetic. Fixes pmodels/mpich#5029

Reported-by: Harmen Stoppels <harmenstoppels@gmail.com>

## Expected Impact

Fix build issue for clang 6 compiler.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

**Comment by raffenet** [comment]
Replacement PR for #5030.

**Comment by raffenet** [comment]
Will cherry-pick this to `3.4.x` once merged.

File: configure.ac [hunk 1]
```diff
@@ -2864,7 +2864,7 @@ fi
 if test "$MPID_NO_FLOAT16" != "yes" ; then
     AC_CACHE_CHECK([whether _Float16 is supported],
     pac_cv_have_float16,[
-    AC_TRY_COMPILE(,[_Float16 a;],
+    AC_TRY_COMPILE(,[_Float16 a=1;_Float16 b=1;_Float16 c=a+b;],
```
Comment (hzhou) [inline configure.ac:2867]: Can we add a space after the semicolons?

**Review by hzhou — APPROVED** [review]
Just a minor readability comment

**Comment by raffenet** [comment]
So testing this manually in a Ubuntu 20.04 container with `apt-get install clang-6.0` doesn't seem to solve the problem. I also tried converting the test to AC_TRY_LINK, but still doesn't work.

**Comment by hzhou** [comment]
> So testing this manually in a Ubuntu 20.04 container with `apt-get install clang-6.0` doesn't seem to solve the problem. I also tried converting the test to AC_TRY_LINK, but still doesn't work.

Can you reproduce the example in https://github.com/pmodels/mpich/pull/5030#issue-562499035 manually?

**Comment by raffenet** [comment]
> > So testing this manually in a Ubuntu 20.04 container with `apt-get install clang-6.0` doesn't seem to solve the problem. I also tried converting the test to AC_TRY_LINK, but still doesn't work.
> 
> Can you reproduce the example in [#5030 (comment)](https://github.com/pmodels/mpich/pull/5030#issue-562499035) manually?

I can. Let me try again and make sure I did everything right...

**Comment by raffenet** [comment]
> > > So testing this manually in a Ubuntu 20.04 container with `apt-get install clang-6.0` doesn't seem to solve the problem. I also tried converting the test to AC_TRY_LINK, but still doesn't work.
> > 
> > 
> > Can you reproduce the example in [#5030 (comment)](https://github.com/pmodels/mpich/pull/5030#issue-562499035) manually?
> 
> I can. Let me try again and make sure I did everything right...

So the conftest is being compiled/linked with `-O2` optimization level by configure, which seems to bypass the symbols used in a default compile/link. I can reproduce with the test program:
```
root@a60846e715bf:~/mpich# clang-6.0 test2.c
/usr/bin/ld: /tmp/test2-61f627.o: in function `main':
test2.c:(.text+0x19): undefined reference to `__gnu_h2f_ieee'
/usr/bin/ld: test2.c:(.text+0x27): undefined reference to `__gnu_h2f_ieee'
/usr/bin/ld: test2.c:(.text+0x38): undefined reference to `__gnu_f2h_ieee'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
root@a60846e715bf:~/mpich# clang-6.0 -O2 test2.c
root@a60846e715bf:~/mpich#
```

**Comment by hzhou** [comment]
I see. How about 1. set CFLAGS=-O0? 2. `return (int) (a + b);`?

**Comment by haampie** [comment]
@hzhou that gets compiled to
```
mov     eax, 2
        ret
```

Otherwise

```
#include <stdlib.h>

int main() {
    _Float16 a = rand();
    _Float16 b = rand();
    return a + b;
}
```

works.

But maybe better is not to rely on compilation tests and just add a flag `--with-fp16`? Support is close to non-existent anyways. On x86_64 there's exactly one instruction for it and that is conversion to float.

**Comment by hzhou** [comment]
> But maybe better is not to rely on compilation tests and just add a flag `--with-fp16`?

I kinda like that, but I am hesitating. While it seems convenient now, it will become unmanageable once we start adding more such options. And if at some point half-float support becomes more common, then it will also become a hassle that one has to explicitly enable it. By that time, no once will remember why some one 10 years ago would make such choice :)

**Comment by raffenet** [comment]
@haampie this version successfully disabled `_Float16` support with the clang 6 and 7 packages on Ubuntu 20.04. Clang 8 gives this message:
```
conftest.c:167:1: error: _Float16 is not supported on this target
```

**Comment by raffenet** [comment]
test:mpich/ch4/most

File: configure.ac [hunk 2]
```diff
@@ -2864,8 +2864,11 @@ fi
 if test "$MPID_NO_FLOAT16" != "yes" ; then
     AC_CACHE_CHECK([whether _Float16 is supported],
     pac_cv_have_float16,[
-    AC_TRY_COMPILE(,[_Float16 a;],
-    pac_cv_have_float16=yes,pac_cv_have_float16=no)])
+    PAC_PUSH_FLAG([CFLAGS])
+    CFLAGS=""
```
Comment (hzhou) [inline configure.ac:2868]: Could you add an explanation comment?

Do you think it is better to set `CFLAGS=-O0`? I am concerned that the compiler may choose a default optimization level.

Comment (raffenet) [reply configure.ac:2868]: > Could you add an explanation comment?

Will do.

> Do you think it is better to set `CFLAGS=-O0`? I am concerned that the compiler may choose a default optimization level.

Possibly. MPICH's setting of the default optimization level is what got us into this mess 😆 . I'll set it to `-O0` in the next push.

**Comment by haampie** [comment]
I thought I had left a comment, but apparently I didn't. According to Godbolt.org adding `volatile` to a and b works as well and doesn't require optimization flags (maybe more portable?)

**Review by hzhou — APPROVED** [review]

**Comment by raffenet** [comment]
> I thought I had left a comment, but apparently I didn't. According to Godbolt.org adding `volatile` to a and b works as well irrespective of optimization flags (maybe more portable?)

Oh, I kind of like that. Basically a standard C way of saying don't run optimizations on this variable.

**Comment by raffenet** [comment]
Output from `clang-7 -O2` with the `volatile` version just pushed.
```
main:                                   # @main
        push    rax
        mov     word ptr [rsp + 6], 15360
        movzx   edi, word ptr [rsp + 6]
        call    __gnu_h2f_ieee
        cvttss2si       eax, xmm0
        pop     rcx
        ret
```

**Comment by raffenet** [comment]
test:mpich/ch4/most

**Review by hzhou — APPROVED** [review]
